### PR TITLE
[metrics] Collect generic metrics from Marketplace using SDK metrics package

### PR DESCRIPTION
**Pull request by varshaprasad96**
Used SDK provided functions to collect generic Prometheus metrics from marketplace.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -10,6 +10,10 @@ required = [
   "sigs.k8s.io/controller-tools/pkg/crd/generator",
 ]
 
+[[override]]
```
Comment (awgreene) [inline Gopkg.toml:13]: Can you confirm with the sdk team that this import is required.

File: cmd/manager/main.go [hunk 2]
```diff
@@ -32,13 +38,20 @@ const (
 	updateNotificationSendWait = time.Duration(10) * time.Minute
 )
 
+var (
+	// Change the below variables to serve metrics on different host or port.
```
Comment (awgreene) [inline cmd/manager/main.go]: These should be constants. These should also have descriptions.

File: cmd/manager/main.go [hunk 3]
```diff
@@ -32,13 +38,20 @@ const (
...
+)
+
 func printVersion() {
 	log.Printf("Go Version: %s", runtime.Version())
 	log.Printf("Go OS/Arch: %s/%s", runtime.GOOS, runtime.GOARCH)
 	log.Printf("operator-sdk Version: %v", sdkVersion.Version)
 }
 
 func main() {
+
```
Comment (awgreene) [inline cmd/manager/main.go]: Please remove this newline.

File: cmd/manager/main.go [hunk 4]
```diff
@@ -58,9 +71,14 @@ func main() {
 	}
 
 	// Create a new Cmd to provide shared dependencies and start components
-	mgr, err := manager.New(cfg, manager.Options{Namespace: namespace})
+	mgr, err := manager.New(cfg, manager.Options{
+		Namespace:          namespace,
+		MapperProvider:     restmapper.NewDynamicRESTMapper,
+		MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
```
Comment (awgreene) [inline cmd/manager/main.go:74]: Please explain what this line does.

File: cmd/manager/main.go [hunk 5]
```diff
@@ -58,9 +71,14 @@ func main() {
...
+	})
 	if err != nil {
-		log.Fatal(err)
+		log.Error(err, "")
```
Comment (awgreene) [inline cmd/manager/main.go]: Please revert this change.

File: cmd/manager/main.go [hunk 6]
```diff
@@ -83,6 +101,37 @@ func main() {
 		exit(err)
 	}
 
+	ctx := context.TODO()
+
+	// Create Service object to expose the metrics port.
+	service, err := metrics.ExposeMetricsPort(ctx, metricsPort)
+	fmt.Printf("Service %v", service)
```
Comment (awgreene) [inline cmd/manager/main.go]: Please use the logger rather than `fmt.Printf`.

Also, please update this log to be more descriptive.

File: cmd/manager/main.go [hunk 7]
```diff
@@ -83,6 +101,37 @@ func main() {
...
+	if err != nil {
+		log.Info("ERROR EXPOSING METRICS")
+		log.Info(err.Error())
```
Comment (awgreene) [inline cmd/manager/main.go]: Please log errors that shouldn't prevent marketplace from running as `log.Error(err)`.

Alternatively, you could use `log.Errorf("Unable to expose metrics port: %v", err)`

File: cmd/manager/main.go [hunk 8]
```diff
@@ -83,6 +101,37 @@ func main() {
...
+	}
+
+	// Create one `ServiceMonitor` per application per namespace.
```
Comment (awgreene) [inline cmd/manager/main.go]: 1. Rather than creating a `ns` variable, you can just use the `namespace` variable found above.
2. This would not work for our upstream distribution given that the operator is deployed in `marketplace` rather than `openshift-marketplace`. The suggestion in `1.` fixes this.
3.  Regarding the commit messages:
> // Create one `ServiceMonitor` per application per namespace.

This should be further down above the `metrics.CreateServiceMonitors(cfg, ns, services)` function. Also I'm not a fan of this comment in general.

> // Change below value to name of the Namespace you want the `ServiceMonitor` to be created in.

In general, don't mention that you can change these values. Mention why they are set to the values you selected.

File: cmd/manager/main.go [hunk 9]
```diff
@@ -83,6 +101,37 @@ func main() {
...
+	// Change below value to name of the Namespace you want the `ServiceMonitor` to be created in.
+	ns := "openshift-marketplace"
+	// Populate below with the Service(s) for which you want to create ServiceMonitors.
+	services := []*v1.Service{}
```
Comment (awgreene) [inline cmd/manager/main.go]: Create this inline in the function call instead.

File: cmd/manager/main.go [hunk 10]
```diff
@@ -83,6 +101,37 @@ func main() {
...
+	services = append(services, service)
+
+	fmt.Printf("Services %v", services)
+
+	// Pass the Service(s) to the helper function so that the service-monitor registers the service.
```
Comment (awgreene) [inline cmd/manager/main.go]: This function creates a `ServiceMonitor` that expose marketplace metrics. Your comment should highlight that.

File: cmd/manager/main.go (hunk 7 above)
Comment (awgreene) [reply cmd/manager/main.go]: This applies in all your if statements that catch errors.

File: manifests/05_role.yaml [hunk 11]
```diff
@@ -66,6 +67,8 @@ rules:
   - apps
   resources:
   - deployments
+  - deployments/finalizers
```
Comment (awgreene) [inline manifests/05_role.yaml:70]: Do we really need all of the verbs for these two resources?

verbs:

- get
- create
- delete
- update
- list
- watch

File: manifests/05_role.yaml [hunk 12]
```diff
@@ -85,3 +88,10 @@ rules:
   - update
   - list
   - watch
+- apiGroups:
+  - monitoring.coreos.com
+  resources:
+  - servicemonitors
+  verbs:
+  - get
+  - create
```
Comment (awgreene) [inline manifests/05_role.yaml]: Please add a newline.

File: manifests/05_role.yaml (hunk 11 above)
Comment (awgreene) [reply manifests/05_role.yaml:70]: This comment applies in all RBAC changes.

File: manifests/10_operator.yaml [hunk 13]
```diff
@@ -31,7 +31,7 @@ spec:
         tolerationSeconds: 120
       containers:
         - name: marketplace-operator
-          image: quay.io/openshift/origin-operator-marketplace:latest
+          image: quay.io/vnarsing/operator-marketplace:sdktest
```
Comment (awgreene) [inline manifests/10_operator.yaml]: Please don't update the image.

File: manifests/10_operator.yaml [hunk 14]
```diff
@@ -57,6 +57,10 @@ spec:
                   fieldPath: metadata.namespace
             - name: OPERATOR_NAME
               value: "marketplace-operator"
+            - name: POD_NAME
```
Comment (awgreene) [inline manifests/10_operator.yaml:60]: Please add these changes to the deployments found in `/manifests`,`/deploy`, and `/deploy/upstream`.

File: manifests/05_role.yaml (hunk 11 above)
Comment (awgreene) [reply manifests/05_role.yaml:70]: If you need help testing this please let me know.

File: Gopkg.toml [hunk 15]
```diff
@@ -10,6 +10,10 @@ required = [
   "sigs.k8s.io/controller-tools/pkg/crd/generator",
```
Comment (awgreene) [inline Gopkg.toml:10]: Please update the commit message to include a message and description and follow the 50/72 rule.

**Review by awgreene — CHANGES_REQUESTED** [review]
Great work @varshaprasad96, a few comments.

File: cmd/manager/main.go (hunk 2 above)
Comment (varshaprasad96) [reply cmd/manager/main.go]: The user can change these values if he/she wants the metrics in different port/path. So, shall I change them to constants, or keep them as variables itself?

Comment (awgreene) [reply cmd/manager/main.go]: This is the marketplace operator - users can't change it because it's controlled by out team.

File: Gopkg.toml (hunk 1 above)
Comment (varshaprasad96) [reply Gopkg.toml:13]: This seems to be required I suppose. Without this when we use metrics collection functions, it gives an error as there is a missing dependency.

File: manifests/05_role.yaml (hunk 11 above)
Comment (varshaprasad96) [reply manifests/05_role.yaml:70]: These permissions do not effect the exposing of metrics. But, they do give the following message in the logs:
```
: Failed to list *v1.Deployment: deployments.apps is forbidden: User "system:serviceaccount:openshift-marketplace:marketplace-operator" cannot list resource "deployments" in API group "apps" in the namespace "openshift-marketplace"
```
The same applies to other two resources also.

**Comment by varshaprasad96** [comment]
@awgreene - I have made the suggested changes. Can you please look into the same.

File: Gopkg.toml (hunk 1 above)
Comment (awgreene) [reply Gopkg.toml:13]: Please address this comment.

File: cmd/manager/main.go (hunk 9 above)
Comment (awgreene) [reply cmd/manager/main.go]: Please address this comment.

File: deploy/upstream/05_role.yaml [hunk 16]
```diff
@@ -76,3 +88,11 @@ rules:
...
   - list
   - watch
+- apiGroups:
+  - monitoring.coreos.com
+  resources:
+  - servicemonitors
+  verbs:
+  - get
+  - create
+
```
Comment (awgreene) [inline deploy/upstream/05_role.yaml]: Something is off here.

File: manifests/05_role.yaml (hunk 11 above)
Comment (awgreene) [reply manifests/05_role.yaml:70]: Have you confirmed that all these RBAC changes are required?

File: deploy/upstream/05_role.yaml [hunk 17]
```diff
@@ -13,12 +13,21 @@ rules:
   - delete
   - update
   - list
+- apiGroups:
```
Comment (awgreene) [inline deploy/upstream/05_role.yaml]: You cannot just copy paste into the upstream folder. Vanilla kube does not define this API Group.

File: deploy/upstream/05_role.yaml [hunk 18]
```diff
@@ -13,12 +13,21 @@ rules:
...
+  - create
+  - get
+  - update
 ---
 kind: Role
 apiVersion: rbac.authorization.k8s.io/v1
 metadata:
   name: marketplace-operator
-  namespace: marketplace
+  namespace: openshift-marketplace
```
Comment (awgreene) [inline deploy/upstream/05_role.yaml]: Marketplace does not run in this namespace on vanilla kube.

File: cmd/manager/main.go [hunk 19]
```diff
@@ -83,6 +98,31 @@ func main() {
...
+	service, err := metrics.ExposeMetricsPort(ctx, metricsPort)
+	if err != nil {
+		log.Errorf("Unable to expose metrics port: %v", err)
+	}
+
+	// The below variable refers to the namespace in which the operator is deployed.
+	ns := "openshift-marketplace"
+	// Creates a list of services which are to be registered with the service monitor.
+	services := []*v1.Service{}
+	log.Info("Services provided for Service Monitor creation", services)
```
Comment (awgreene) [inline cmd/manager/main.go]: Lets remove this.

File: cmd/manager/main.go [hunk 20]
```diff
@@ -83,6 +98,31 @@ func main() {
...
+	// Create a serviceMonitor for the namespace, and register the service created above to it.
+	_, err = metrics.CreateServiceMonitors(cfg, ns, append(services, service))
+	if err != nil {
+		log.Errorf("Error Creating Service Monitor: %v", err)
+	} else {
+		log.Info("Service Monitor created successfully")
+
+	}
+
+	log.Info("Expose-Metrics succeded")
```
Comment (awgreene) [inline cmd/manager/main.go]: Let's remove this.

File: cmd/manager/main.go [hunk 21]
```diff
@@ -83,6 +98,31 @@ func main() {
...
+
+	// The below variable refers to the namespace in which the operator is deployed.
+	ns := "openshift-marketplace"
+	// Creates a list of services which are to be registered with the service monitor.
+	services := []*v1.Service{}
+	log.Info("Services provided for Service Monitor creation", services)
+
+	// Create a serviceMonitor for the namespace, and register the service created above to it.
+	_, err = metrics.CreateServiceMonitors(cfg, ns, append(services, service))
+	if err != nil {
```
Comment (awgreene) [inline cmd/manager/main.go:114]: Only log the err if it exists.

**Review by awgreene — CHANGES_REQUESTED** [review]

File: Gopkg.toml (hunk 1 above)
Comment (varshaprasad96) [reply Gopkg.toml:13]: This seems to be required. So, I have left it as it is.

File: deploy/upstream/05_role.yaml (hunk 17 above)
Comment (varshaprasad96) [reply deploy/upstream/05_role.yaml]: Sorry!. Got it. Changed the defaults accordingly.

File: deploy/upstream/05_role.yaml (hunk 16 above)
Comment (varshaprasad96) [reply deploy/upstream/05_role.yaml]: This was a mistake. Reverted to the old version and made the required changes

File: manifests/05_role.yaml (hunk 11 above)
Comment (varshaprasad96) [reply manifests/05_role.yaml:70]: Yes, I suppose. Or else, for each of it, there is a log message similar to the above

File: Gopkg.toml (hunk 1 above)
Comment (varshaprasad96) [reply Gopkg.toml:13]: Spoke with SDK team, we would require this to be manually added, if we use `metrics package` provided by them.

**Comment by varshaprasad96** [comment]
@awgreene - I have made the changes as suggested. Can you please review.
